### ISSUE TITLE
Idea for new helper method FieldList::dataFieldByType()

### ISSUE THREAD

**Issue by sminnee** [silverstripe/silverstripe-framework#9659]
I've been playing around with type hinting, and forcing the use of /** @var */ is irksome. I'm keen to see if we can provide mechanisms for avoiding it with dynamic type hinting.

```php
$fields = parent::getCMSFields();

/** @var GridField */
$grid = $fields->dataFieldByName('Items');
$header = $grid->getConfig()->getComponentByType(GridFieldSortableHeader::class)
```

I'm wondering in this would be better:

```php
$fields = parent::getCMSFields();
$grid = $fields->dataFieldByType(GridField::class, 'Items');
$header = $grid->getConfig()->getComponentByType(GridFieldSortableHeader::class)
```

The dataFieldByType() method would do 3 things:
 * Allow type hinting for VS Code and PHP form to know the return type
 * Throw an error if the given field was _not_ of the given type, introducing some more type safety
 * The 2nd argument could be optional, in which case it would just return the first grid field. For stuff like modeladmin customisation this would be handy - there's always a single grid field in a modeladmin, but its class changes.

The implementation would be something like:

```php
function dataFieldByType(string $class, string $name = null)
{
  if ($name  !== null) {
    $field = $this->dataFieldByName($name);
    if (!($field instanced $name)) throw \InvalidArgumentException("Field $name is not of class $class");
    return $field;
  } else {
    foreach ($this->dataFields() as $field) { if ($field instanced $class) { return $field); }
    if (!($field instanced $name)) throw \InvalidArgumentException("No field of class $class found");
  }
}
```

By adding this as a new method rather than trying to modify the behaviour of existing methods, backward-compat is easy

Before doing a PR I was keen to get feedback on how useful this kind of type-safety-support method would be to others.

---

If this approach was useful, a similar method could be provided `Injector::getWithType($serviceName, $type)`. Although in many cases Injector::get($serviceName) will return an instance of service name, that's not always the case.

For example, this currently exists in Director's routing:
```php
/** @var RequestHandler $controllerObj */
$controllerObj = Injector::inst()->create($arguments['Controller']);
```

If we included something like this, then it could throw an error if the argument returned something other than a RequestHandler.
```php
$controllerObj = Injector::inst()->createWithType($arguments['Controller'], RequestHandler::class);
```

**Comment by michalkleiner** [comment]
From the suggested implementation (I get it's an indication only) I'm not sure I understand how would the `dataFieldByType` method get the type and how would it be useful for static analysis.

**Comment by sminnee** [comment]
You'd need to populate .phpstorm.meta.php as well (which also works in VS Code) - see #9658. Other static analysers could be configured also, e.g. with a PR to https://github.com/symbiote/silverstripe-phpstan.

**Comment by michalkleiner** [comment]
Right, so on its own it won't be of much help without other helpers/tooling understanding the mechanism of it.

**Comment by sminnee** [comment]
Yeah basically it makes the configuration of those other helpers possible.

**Comment by sminnee** [comment]
It's also worth noting that you could do something similar with injector, where another method let you define the expected class or interface and throw an error if that was bad.

**Comment by sminnee** [comment]
It looks like phpstan could be amended to work with such methods merely by changing the docblocks https://phpstan.org/blog/generics-in-php-using-phpdocs

**Comment by robbieaverill** [comment]
What happens if you call `$fields->dataFieldByType(TextareaField::class, 'Description')` but an extension has changed its type to another type of field?

**Comment by sminnee** [comment]
> What happens if you call $fields->dataFieldByType(TextareaField::class, 'Description') but an extension has changed its type to another type of field?

If it's replaced with a subclass of TextareaField then in either case is fine (the check would be an `instanceof` check which subclasses pass)

If it's a completely different check:

 * Status quo: your code that expected a TextareaField will run with some different class. Maybe it works, maybe it will break, maybe the break won't be noticed until production, maybe the break will cause data loss

 * With my proposed change: the code will break immediately saying "TextareaField expected but DropdownField found"

In other words, I'm arguing that an immediate failure is an improvement on the status-quo.

**Comment by michalkleiner** [comment]
Well, if the current approach stays available too, meaning you'll be able to get a field by its name regardless of its type to for example add a description to it (where you don't care about its type), then all good.
If a dev chooses to have a bit stricter control e.g. to always get a WYSIWYG to apply config to it, this might be a good option.

**Comment by sminnee** [comment]
Yes, `dataFieldByName()` will remain as an option. If we were to use `dataFieldByType()` in core code or widely-used modules, this may have an impact to sites that have customised built-in fields. As such I would recommend using `dataFieldByType()` only in cases where the correct type is necessary for subsequent code to function correctly. In particular, I would advocate using phpstan to identify overly loose type assumptions and fix them with dataFieldByType() to guard against currently unprotected error cases.

**Comment by emteknetnz** [comment]
> the use of /** @var */ is irksome

It is pretty ugly syntax, though that is basically the "proper" way to "declare" a type in PHP.  

If you we want to do more type check at runtime, possibly we write a static utility functions somewhere and use that instead of having to change a bunch of function to equivalent functions that also include a type check. e.g.

```php
class MyUtilityThing {
  public static confirmClass($object, $class):void {
    if (!($object instanceof $class)) throw \InvalidArgumentException("Object $object is not of class $class");
  }
}

...
$grid = $fields->dataFieldByName('Items');
MyUtilityThing::confirmClass($grid, GridField::class);
```

**Comment by sminnee** [comment]
I guess you could do that, but personally I'd prefer a type-aware API. In particular:

 * it combines the runtime check and static analysis hint into the same construct. (I'm not sure if the assert-style utility method you propose can feed appropriate metadata to PHPStorm)
 * it makes it harder to forget to do the type (if you call the method, it's a required arg)
 * it's easier to chain for brevity, where that's of use
 * static methods on a separate utility class add more cross-subsystem coupling

Less importantly, it adds the edge-case feature of choosing a field without a name (which is basically for getting the grid field in a model admin's root form)

Happy to hear other opinions though, if I'm in the minority here @silverstripe/core-team

**Comment by robbieaverill** [comment]
I have no problem at all with `/** @var TypeHere $varName */` syntax.

Ideally we start implementing more type hints as we built new PHP 7 compatible APIs, then we need less of the PSR-5/19 annotations.

**Comment by emteknetnz** [comment]
I'm more with Robbie here, I'm used to ugly syntax and I'd prefer to not create a non-native type checking implementation

That said, you could do it this way instead

```php
trait MyTrait
{
    // possibly misleading naming as it's just asserting the type instead of casting it ...
    public function asType(string $type) {
       $thisClass = get_class($this);
        if (!($this instanceof $class)) throw \InvalidArgumentException("Object $thisClass is not of class $class");
        return $this;
    }
}

class DataObject / GridField / DataList ...
  use MyTrait

$grid = $fields->dataFieldByName('Items')->asType(GridField::class)->doSomething();
```

Possibly you could then do something with MyTrait::asType in .phpstrom.meta.php?
https://github.com/silverstripe/silverstripe-framework/issues/9658

**Comment by sminnee** [comment]
> Ideally we start implementing more type hints as we built new PHP 7 compatible APIs, then we need less of the PSR-5/19 annotations.

Yeah, nah, that's not possible without something like what's suggested here. Because `dataFieldByName` can only be hinted to `FormField`, but very frequently will be assumed to be, e.g. `GridField` in the code that follows.

PHP's "real" typing lacks the sophistication for something like dataFieldByType / asType to work, but phpstan and phpstorm both do.

The trait approach is interesting, Steve. I'd probably call it "assertType" at that point, but you could build that trait and add it to FormField.

**Comment by dhensby** [comment]
I'd be happy to add a utility method like this, as I think it has a purpose beyond purely helping out with type hinting, but I'm not completely convinced.

Are we creating more upgrade work for ourselves later if we start specifying types that *could* be subject to change? SilverStripe hasn't typically been strict with types; is that something we want to move more towards?

**Comment by sminnee** [comment]
If you have written code that fetches a form field and calls ->getConfig()->getComponentByType() on the result, then if the field returned is anything other than a GridField, you're going to have a bad time. So forcing it to be GridField doesn't make things harder.

The point is to use this helper sparingly, only to ensure that the method calls you make align with the type being asserted. Which phpstan or PHPstorm will warn you about if the typing is too loose.

**Comment by sminnee** [comment]
Alternatively we could use this:

```php
$fields = parent::getCMSFields();
$grid = $fields->dataFieldByName('Items');
assert($grid instanceof GridField);
$header = $grid->getConfig()->getComponentByType(GridFieldSortableHeader::class);
```

Compared to the manual type-hint the code looks similar:

```php
$fields = parent::getCMSFields();
/** @var GridField */
$grid = $fields->dataFieldByName('Items');
$header = $grid->getConfig()->getComponentByType(GridFieldSortableHeader::class);
```

Key advantage is that it will fail at runtime as well as pick up a type error.

 * It works in PHPStAn https://phpstan.org/r/9b405a35-5203-45f7-9cd0-2e611d9e2331
 * It also works in PHPStorm and VS Code type hinting

Since it needs no new code, perhaps that's better? Use assert(... instanceof ...) in favour of `@var` to get both runtime and lint-time type awareness?

**Comment by sminnee** [comment]
I'm going to close this I think, assert() is less invasive.
